### ISSUE TITLE
Psedet for other detectors

### ISSUE THREAD

**Issue by pILLOW-1** [wang-qiuchen/PseDet#5]
Hi, great work!
I am trying to implement psedet for other detectors, and currently I choose DINO. However, I find that the performance is not ideal. For example, I   replace the gfl with dino using the default dino setting from mmdet, and keep the psedet setting the same as gfl. I train the dino model under 40+40 setting for 12 epochs. When the training reaches the 4th epoch, the result is weird as follows,

### AP_old = 1.6; AP_new: 37.0; AP_total: 19.3

The gfl has the following result,

### AP_old = 38.0; AP_new: 30.6; AP_total: 34.3

I think the dino result does not make sense, especially for old classes, but I don't know what's wrong. Could you please give some advice? Thanks a lot!

**Comment by wang-qiuchen** [comment]
How about the performance of the model in stage2 before epoch_0; 

maybe you can check https://github.com/wang-qiuchen/PseDet/blob/93d64b9074028f7527a3b3e72edf479fbc394bb3/coco_cfg/train_stage_40_20_20/gfl_r50_1x_422_s2_80cls.py#L189C16-L189C32 about the loading weight module?

**Comment by pILLOW-1** [comment]
I find the answer. CLLoadCheckpoint is tailored for GFL, but DINO has a different architecture. So, when I load this hook, it will actually be skipped! I really forgot that! I will change it to DINO version and train again. I think this is the root of this problem. Thanks a lot!

**Comment by pILLOW-1** [comment]
I try the DINO under 40+40 and 70+10 settings. The mAP is 30.8 and 35.7, respectively. Does that make sense?
Below are my custom checkpoint loading hook and config files.

checkpoint loading hook: 
`
import torch
from mmengine.hooks import Hook
from mmengine.runner import Runner
from mmengine.registry import HOOKS
import pdb

@HOOKS.register_module()
class LoadPartialClsHeadHook(Hook):
    def __init__(self, num_old_classes, det_classes, frozen=False,  head_name='bbox_head.cls_branches', label_embedding_name='dn_query_generator.label_embedding'):
        self.num_old_classes = num_old_classes
        self.head_name = head_name
        self.frozen = frozen
        self.label_embedding_name = label_embedding_name
        self.det_classes = det_classes
        
    def after_load_checkpoint(self, runner: Runner, checkpoint: dict) -> None:
        model = runner.model
        state_dict = checkpoint['state_dict']

        runner.model.module.bbox_head.det_classes = self.det_classes

        def get_attr_by_path(obj, attr_path):
            if hasattr(obj, 'module'):
                real_model = obj.module  
            else:
                real_model = obj         

            for attr in attr_path.split('.'):
                real_model = getattr(real_model, attr)
            return real_model
        
        # tailored for dino cls branches, because there are six levels
        for i in range(7):
            head_name = self.head_name + f'.{str(i)}'
            fc_cls = get_attr_by_path(model, head_name)
            old_weight = state_dict.get(f'{head_name}.weight', None)
            old_bias = state_dict.get(f'{head_name}.bias', None)

            if old_weight is None and old_bias is None:
                print(f'[Hook] Warning: Could not find {head_name} in checkpoint.')
                return

            num_loaded = self.num_old_classes
            print(f'[Hook] Loading first {num_loaded} classes from {head_name}.')

            fc_cls.weight.data[:num_loaded] = old_weight[:num_loaded]
            fc_cls.bias.data[:num_loaded] = old_bias[:num_loaded]
            fc_cls.weight.requires_grad = True
            fc_cls.bias.requires_grad = True

            if self.frozen:

                fc_cls.weight.register_hook(self.freeze_partial_grad)
                fc_cls.bias.register_hook(self.freeze_partial_grad)

            print(f'[Hook] Successfully loaded weights and registered freeze hook for first {num_loaded} classes for cls_branch.')
        
        # label_embedding
        label_embedding_name = self.label_embedding_name
        label_embedding = get_attr_by_path(model, label_embedding_name)
        old_weight = state_dict.get(f'{label_embedding_name}.weight', None)

        if old_weight is None:
            print(f'[Hook] Warning: Could not find {label_embedding_name} in checkpoint.')
            return

        num_loaded = self.num_old_classes
        print(f'[Hook] Loading first {num_loaded} classes from {label_embedding_name}.')

        label_embedding.weight.data[:num_loaded] = old_weight[:num_loaded]
        label_embedding.weight.requires_grad = True

        if self.frozen:
            label_embedding.weight.register_hook(self.freeze_partial_grad)

        print(f'[Hook] Successfully loaded weights and registered freeze hook for first {num_loaded} classes for query&label embedding.')
    
    def freeze_partial_grad(self, grad):
        frozen = torch.zeros_like(grad[:self.num_old_classes])
        return torch.cat([frozen, grad[self.num_old_classes:]], dim=0)
    
    def freeze_full_grad(self, grad):
        frozen = torch.zeros_like(grad)
        return frozen
`

70+10 config file:
`import platform,os
data_root = '/root/workspace/d2h97tc7v38s73b4dfig/coco/'
checkpoint_path = '/root/workspace/d2h97tc7v38s73b4dfig/PseDet/checkpoint/resnet50-0676ba61.pth'

load_from = '/root/workspace/d2h97tc7v38s73b4dfig/PseDet/work_dirs/dino_r50_1x_s0_70cls/epoch_12.pth'
pseudo_file = './work_dirs/dino_pseudo_kmeans_70_psedet.pkl'

stage_index = 1
num_classes = 80
classes_groups=[70,10]

model = dict(
    type='DINO',
    num_queries=900,  # num_matching_queries
    with_box_refine=True,
    as_two_stage=True,
    data_preprocessor=dict(
        type='DetDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True,
        pad_size_divisor=1),
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=False),
        norm_eval=True,
        style='pytorch',
        init_cfg=dict(type='Pretrained', checkpoint=checkpoint_path)),
    neck=dict(
        type='ChannelMapper',
        in_channels=[512, 1024, 2048],
        kernel_size=1,
        out_channels=256,
        act_cfg=None,
        norm_cfg=dict(type='GN', num_groups=32),
        num_outs=4),
    encoder=dict(
        num_layers=6,
        layer_cfg=dict(
            self_attn_cfg=dict(embed_dims=256, num_levels=4,
                               dropout=0.0),  # 0.1 for DeformDETR
            ffn_cfg=dict(
                embed_dims=256,
                feedforward_channels=2048,  # 1024 for DeformDETR
                ffn_drop=0.0))),  # 0.1 for DeformDETR
    decoder=dict(
        num_layers=6,
        return_intermediate=True,
        layer_cfg=dict(
            self_attn_cfg=dict(embed_dims=256, num_heads=8,
                               dropout=0.0),  # 0.1 for DeformDETR
            cross_attn_cfg=dict(embed_dims=256, num_levels=4,
                                dropout=0.0),  # 0.1 for DeformDETR
            ffn_cfg=dict(
                embed_dims=256,
                feedforward_channels=2048,  # 1024 for DeformDETR
                ffn_drop=0.0)),  # 0.1 for DeformDETR
        post_norm_cfg=None),
    positional_encoding=dict(
        num_feats=128,
        normalize=True,
        offset=0.0,  # -0.5 for DeformDETR
        temperature=20),  # 10000 for DeformDETR
    bbox_head=dict(
        type='DINOHead',
        num_classes=num_classes,
        sync_cls_avg_factor=True,
        loss_cls=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),  # 2.0 in DeformDETR
        loss_bbox=dict(type='L1Loss', loss_weight=5.0),
        loss_iou=dict(type='GIoULoss', loss_weight=2.0)),
    dn_cfg=dict(  # TODO: Move to model.train_cfg ?
        label_noise_scale=0.5,
        box_noise_scale=1.0,  # 0.4 for DN-DETR
        group_cfg=dict(dynamic=True, num_groups=None,
                       num_dn_queries=100)),  # TODO: half num_dn_queries
    # training and testing settings
    train_cfg=dict(
        assigner=dict(
            type='HungarianAssigner',
            match_costs=[
                dict(type='FocalLossCost', weight=2.0),
                dict(type='BBoxL1Cost', weight=5.0, box_format='xywh'),
                dict(type='IoUCost', iou_mode='giou', weight=2.0)
            ])),
    test_cfg=dict(max_per_img=300),
)  # 100 for DeformDETR

# optimizer
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(
        type='AdamW',
        lr=0.0001,  # 0.0002 for DeformDETR
        weight_decay=0.0001),
    clip_grad=dict(max_norm=0.1, norm_type=2),
    paramwise_cfg=dict(custom_keys={'backbone': dict(lr_mult=0.1)})
)  # custom_keys contains sampling_offsets and reference_points in DeformDETR  # noqa

# learning policy
max_epochs = 12


# dataset settings
# dataset_type = 'CocoDataset'
dataset_type = 'CocoDatasetCLMultiStage'

backend_args = None

# train_pipeline, NOTE the img_scale and the Pad's size_divisor is different
# from the default setting in mmdet.
train_pipeline = [
    dict(type='LoadImageFromFile', backend_args=backend_args),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='RandomFlip', prob=0.5),
    dict(
        type='RandomChoice',
        transforms=[
            [
                dict(
                    type='RandomChoiceResize',
                    scales=[(480, 1333), (512, 1333), (544, 1333), (576, 1333),
                            (608, 1333), (640, 1333), (672, 1333), (704, 1333),
                            (736, 1333), (768, 1333), (800, 1333)],
                    keep_ratio=True)
            ],
            [
                dict(
                    type='RandomChoiceResize',
                    # The radio of all image in train dataset < 7
                    # follow the original implement
                    scales=[(400, 4200), (500, 4200), (600, 4200)],
                    keep_ratio=True),
                dict(
                    type='RandomCrop',
                    crop_type='absolute_range',
                    crop_size=(384, 600),
                    allow_negative_crop=True),
                dict(
                    type='RandomChoiceResize',
                    scales=[(480, 1333), (512, 1333), (544, 1333), (576, 1333),
                            (608, 1333), (640, 1333), (672, 1333), (704, 1333),
                            (736, 1333), (768, 1333), (800, 1333)],
                    keep_ratio=True)
            ]
        ]),
    dict(type='PackDetInputs')
]

test_pipeline = [
    dict(type='LoadImageFromFile', backend_args=backend_args),
    dict(type='Resize', scale=(1333, 800), keep_ratio=True),
    # If you don't have a gt annotation, delete the pipeline
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PackDetInputs',
        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                   'scale_factor'))
]
train_dataloader = dict(
    batch_size=8,
    num_workers=2,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    batch_sampler=dict(type='AspectRatioBatchSampler'),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='annotations/instances_train2017.json',
        pseudo_file=pseudo_file,
        data_prefix=dict(img=data_root + 'images/train2017/'),
        filter_cfg=dict(filter_empty_gt=True, min_size=32,
                classes_groups=classes_groups,
                class_spilt=True,
                stage_index = stage_index),
        pipeline=train_pipeline,
        backend_args=backend_args))
val_dataloader = dict(
    batch_size=2,
    num_workers=2,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='annotations/instances_val2017.json',
        data_prefix=dict(img=data_root+'images/val2017/'),
        test_mode=True,
        filter_cfg=dict(equally_subset=False, class_spilt=False),
        pipeline=test_pipeline,
        backend_args=backend_args))
test_dataloader = val_dataloader

val_evaluator = [
    # dict(
    #     type='CocoMetric',
    #     # ann_file=os.path.join(data_root,'annotations/instances_val2017.json'),
    #     metric='bbox',
    #     format_only=False,
    #     backend_args=backend_args),
    dict(
        type='CocoMetricCLMultiStage',
        class_group=classes_groups,
        # ann_file=os.path.join(data_root,'annotations/instances_val2017.json'),
        metric='bbox',
        format_only=False,
        backend_args=backend_args)
]
test_evaluator = val_evaluator


# training schedule for 1x
train_cfg = dict(type='EpochBasedTrainLoop', max_epochs=12, val_interval=1)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')

param_scheduler = [
    dict(
        type='MultiStepLR',
        begin=0,
        end=max_epochs,
        by_epoch=True,
        milestones=[11],
        gamma=0.1)
]

# NOTE: `auto_scale_lr` is for automatically scaling LR,
# USER SHOULD NOT CHANGE ITS VALUES.
# base_batch_size = (8 GPUs) x (2 samples per GPU)
auto_scale_lr = dict(base_batch_size=16)

default_scope = 'mmdet'

default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=50),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=1, max_keep_ckpts=1),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    visualization=dict(type='DetVisualizationHook'))

custom_hooks = [
    dict(type='LoadPartialClsHeadHook', num_old_classes=70, det_classes=80)
]

env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'),
)

vis_backends = [dict(type='LocalVisBackend')]
visualizer = dict(
    type='DetLocalVisualizer', vis_backends=vis_backends, name='visualizer')
log_processor = dict(type='LogProcessor', window_size=50, by_epoch=True)

log_level = 'INFO'

resume = False
`

**Comment by pILLOW-1** [comment]
The DINO performance is 30.8(40+40) and 35.7(70+10). Does that make sense?
Below are my custom checkpoint loading hook file and config file(70+10). 

[checkpoint_load.py](https://github.com/user-attachments/files/22437631/checkpoint_load.py)
[dino_70_10cls.py](https://github.com/user-attachments/files/22437630/dino_70_10cls.py)

**Comment by pILLOW-1** [comment]
Here are more specific metric details: 
The 1st stage: 
    AP_old: 49.6(40+40)、48.5(70+10)
The 2nd stage:
    70+10: AP_old=34.7, AP_new=42.2, AP_all=35.7
    40+40: AP_old=21.8, AP_new=39.9, AP_all=30.8

**Comment by wang-qiuchen** [comment]
Sorry, I have no idea about it because I haven’t tried it on DINO. However, since the number of labels across categories in the COCO dataset is imbalanced, you might want to try using different seeds when sampling the dataset.

**Comment by pILLOW-1** [comment]
Could you please release the config files for Deformable DETR? I have tried to reimplement, but the performance is 35.7(70+10 setting), which is not 44.7 reported in the paper. Is there something wrong? The following files are my config files. The total batch_size is 32.

[deform_detr_70_10cls.py](https://github.com/user-attachments/files/22582995/deform_detr_70_10cls.py)
[deform_detr_infer_classes_70.py](https://github.com/user-attachments/files/22582998/deform_detr_infer_classes_70.py)
[deform_detr_r50_1x_s0_70cls.py](https://github.com/user-attachments/files/22582997/deform_detr_r50_1x_s0_70cls.py)
[PartialClsHeadHook.py](https://github.com/user-attachments/files/22582996/PartialClsHeadHook.py)

**Comment by wang-qiuchen** [comment]
Of course，i will update later. I just use the official setting of deformable detr. 
And for the config file, you should try implement on two-stage version.

**Comment by pILLOW-1** [comment]
Oh, this is really a detail. I see. Thanks!

**Comment by pILLOW-1** [comment]
I have tried the two-stage version, but the performance is 39.9(70+10) and 37.9(40+40), which is still different from the result in the paper.

**Comment by wang-qiuchen** [comment]
show me your newest config please

**Comment by pILLOW-1** [comment]
[deform_detr_70_10cls.py](https://github.com/user-attachments/files/22791641/deform_detr_70_10cls.py)
This is the updated config file for 70+10 setting. I set 'with_box_refine' and 'as_two_stage' to True.

**Comment by wang-qiuchen** [comment]
can you reproduce the upperbound with your setting?

**Comment by wang-qiuchen** [comment]
hi,
I noticed that your batch size is set to 8, and I'm not sure if you're training with 4 GPUs ? And we suggest aligning with the official setup, or train on 8 gpus with bs=4 may help.

The `milestones` parameter for `MultiStepLR` is set to 11, which might affect the training, but I'm not sure.

Maybe you could try reproducing the official upperbound on your device, then proceed with implementing psedet with your hyper parameters.

**Comment by pILLOW-1** [comment]
Yes, I use bs=8 with 4 GPUs. MultiStepLR is set to 40 as in official setting. Actually, the most content of the config is copied from the official deform_detr config. Is my config the same as yours?

Reproducing the official upperbound takes too long, but I can give you this result:
The AP for the first stage under 70+10 setting is 46. Is that the same as your result?

**Comment by wang-qiuchen** [comment]
> Yes, I use bs=8 with 4 GPUs. MultiStepLR is set to 40 as in official setting. Actually, the most content of the config is copied from the official deform_detr config. Is my config the same as yours?
> 
> Reproducing the official upperbound takes too long, but I can give you this result: The AP for the first stage under 70+10 setting is 46. Is that the same as your result?

What about the 40+40?

**Comment by pILLOW-1** [comment]
The AP for the first stage under the 40+40 setting is also 46.

**Comment by wang-qiuchen** [comment]
> The AP for the first stage under the 40+40 setting is also 46.

May i ask the performance of first 40 categories before step 0 in stage 2 under 40+40?
and the performance of first 40 categories in stage 2 after training under 40+40?

**Comment by pILLOW-1** [comment]
How to evaluate the performance of 'before step 0 in stage 2'? I don't quite understand. 
After training for 1 epoch in stage 2: AP_old = 36.9
After training in stage 2: AP_old = 36.0

**Comment by wang-qiuchen** [comment]
> How to evaluate the performance of 'before step 0 in stage 2'? I don't quite understand. After training for 1 epoch in stage 2: AP_old = 36.9 After training in stage 2: AP_old = 36.0

1. set rl = 1e-10 (or a small value), eval at step iteration_1

2. change the mmdet train code

I think it's important to check whether the load checkpoint function works properly.

and implement calibration supervision in the classification loss
